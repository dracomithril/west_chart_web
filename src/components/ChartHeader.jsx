/**
 * Created by XKTR67 on 5/25/2017.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { Button, ButtonGroup, Label, OverlayTrigger, Popover } from 'react-bootstrap';
import SongsPerDay from './SongsPerDay';
import FilteringOptions from './FilteringOptions';
import PickYourDate from './PickYourDate';
import './components.css';
import action_types from './../reducers/action_types';
import utils, { filterChart } from '../utils/utils';

const ChartButtons = ({ since, until, last_update, onGetDataClick, onQuickSummaryClick }) => {
  const options = { weekday: 'short', month: '2-digit', day: 'numeric' };
  return (
    <div className="chartButtons">
      <OverlayTrigger
        trigger={['hover', 'focus']}
        placement="top"
        overlay={
          <Popover id="update_info">
            <span>since: </span>
            <Label bsStyle="success">
              {since !== '' ? new Date(since).toLocaleDateString('pl-PL', options) : 'null'}
            </Label>
            <span> to </span>
            <Label bsStyle="danger">
              {until !== '' ? new Date(until).toLocaleDateString('pl-PL', options) : 'null'}
            </Label>
            <br />
            <small id="updateDate">{` Last update: ${new Date(last_update).toLocaleString(
              'pl-PL',
            )}`}</small>
          </Popover>
        }
      >
        <div>
          <ButtonGroup vertical>
            <Button id="updateChartB" onClick={onGetDataClick} bsStyle="primary">
              Update
            </Button>
            <Button id="quickSummary" onClick={onQuickSummaryClick} bsStyle="success">
              Quick summary
            </Button>
          </ButtonGroup>
        </div>
      </OverlayTrigger>
    </div>
  );
};
ChartButtons.propTypes = {
  since: PropTypes.number,
  until: PropTypes.number,
  last_update: PropTypes.string,
  onGetDataClick: PropTypes.func,
  onQuickSummaryClick: PropTypes.func,
};

function selectedItem(id) {
  return { type: action_types.TOGGLE_SELECTED, id, checked: true };
}

export default class ChartHeader extends React.Component {
  quickSummary = () => {
    const { store } = this.context;
    utils
      .UpdateChart(store)
      .then(() => {
        const { chart, filters, until, songs_per_day } = store.getState();
        store.dispatch({ type: action_types.TOGGLE_FILTER, id: 'create', checked: true });
        store.dispatch({ type: action_types.UPDATE_DAYS, id: 'create', value: 5 });
        const { view_chart } =
          chart.length > 0 ? filterChart(chart, filters, until, songs_per_day) : { view_chart: [] };
        view_chart.forEach(({ id }) => {
          store.dispatch(selectedItem(id));
        });
        // store.dispatch({type: action_types.TOGGLE_ALL});
        return Promise.resolve();
      })
      .then(() => {
        const start_bt = document.getElementById('start_sp_button');
        start_bt.click();
        const gen_bt = document.getElementById('genName_sp_button');
        gen_bt.click();
        const tab = document.getElementById('chart_playlist_tab');
        tab.click();
      });
  };

  render() {
    const { store } = this.context;
    const { since, until, last_update, songs_per_day } = store.getState();
    const { error_days } = this.props;
    return (
      <div className="chartHeader">
        <div className="dock1">
          <SongsPerDay
            error_days={error_days}
            songs_per_day={songs_per_day}
            onDaysChange={days =>
              store.dispatch({
                type: 'UPDATE_SONGS_PER_DAY',
                days,
              })
            }
          />
          <PickYourDate />
        </div>
        <ChartButtons
          since={since}
          last_update={last_update}
          until={until}
          onQuickSummaryClick={this.quickSummary}
          onGetDataClick={() => utils.UpdateChart(store)}
        />
        <FilteringOptions />
      </div>
    );
  }
}
ChartHeader.contextTypes = {
  store: PropTypes.object,
};
ChartHeader.propTypes = {
  error_days: PropTypes.array,
};
