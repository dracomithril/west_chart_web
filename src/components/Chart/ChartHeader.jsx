/**
 * Created by XKTR67 on 5/25/2017.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { Button, ButtonGroup } from 'react-bootstrap';
import SongsPerDay from './SongsPerDay';
import FilteringOptions from '../FilteringOptions/index';
import PickYourDate from './PickYourDate';
import '../components.css';
import './chart.css';
import { action_types } from './../../reducers/action_types';
import { UpdateChart } from '../../utils/utils';

const ChartButtons = ({ onGetDataClick, onQuickSummaryClick }) => (
  <div className="chart-header__group-buttons">
    <ButtonGroup vertical>
      <Button id="updateChartB" onClick={onGetDataClick} bsStyle="primary">
        Update
      </Button>
      <Button id="quickSummary" onClick={onQuickSummaryClick} bsStyle="success">
        Quick summary
      </Button>
    </ButtonGroup>
  </div>
);
ChartButtons.propTypes = {
  onGetDataClick: PropTypes.func,
  onQuickSummaryClick: PropTypes.func,
};

function selectedItem(id) {
  return { type: action_types.TOGGLE_SELECTED, id, checked: true };
}

export default class ChartHeader extends React.Component {
  quickSummary = () => {
    const { store } = this.context;
    UpdateChart(store)
      .then(() => {
        store.dispatch({ type: action_types.TOGGLE_FILTER, id: 'create', checked: true });
        store.dispatch({ type: action_types.UPDATE_DAYS, id: 'create', value: 5 });
        this.props.view_chart.forEach(({ id }) => {
          store.dispatch(selectedItem(id));
        });
        // store.dispatch({type: action_types.TOGGLE_ALL});
        return Promise.resolve();
      })
      .then(() => {
        const start_bt = document.getElementById('start_sp_button');
        start_bt.click();
        const gen_bt = document.getElementById('genName_sp_button');
        gen_bt.click();
        const tab = document.getElementById('chart_playlist_tab');
        tab.click();
      });
  };

  render() {
    const { store } = this.context;
    const { enable_until, songs_per_day, start_date, show_last } = store.getState();
    const { error_days } = this.props;
    const onDateChange = date => {
      store.dispatch({ type: action_types.UPDATE_START_TIME, date });
    };
    const onDaysChange = target => {
      store.dispatch({ type: action_types.UPDATE_SHOW_LAST, days: Number(target.value) });
    };

    const onChange = target =>
      store.dispatch({
        type: action_types.TOGGLE_ENABLE_UNTIL,
        checked: target.checked,
      });
    return (
      <div className="chart-header">
        <div className="chart-header__left_dock">
          <PickYourDate
            checked={enable_until}
            start_date={start_date}
            show_last={show_last}
            onChange={onChange}
            onDaysChange={onDaysChange}
            onDateChange={onDateChange}
          />
          <SongsPerDay
            error_days={error_days}
            songs_per_day={songs_per_day}
            onDaysChange={days =>
              store.dispatch({
                type: 'UPDATE_SONGS_PER_DAY',
                days,
              })
            }
          />
        </div>
        <ChartButtons
          onQuickSummaryClick={this.quickSummary}
          onGetDataClick={() => UpdateChart(store)}
        />
        <FilteringOptions />
      </div>
    );
  }
}
ChartHeader.contextTypes = {
  store: PropTypes.object,
};
ChartHeader.propTypes = {
  error_days: PropTypes.array,
  view_chart: PropTypes.array,
};
