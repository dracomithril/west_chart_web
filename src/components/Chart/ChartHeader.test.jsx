// @flow
import React from 'react';
import PropTypes from 'prop-types';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallowToJson } from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import { containerWithStyles as ChartHeader } from './ChartHeader';

Enzyme.configure({ adapter: new Adapter() });
const initial_state = require('../../___data___/initial_state.json');

const mockStore = configureMockStore([]);
describe('<ChartHeader/>', () => {
  it('renders without crashing ChartHeader', () => {
    const store = mockStore(initial_state);
    const wrapper = shallow(
      <ChartHeader errorDays={[]} />, {
        context: { store },
        childContextTypes: { store: PropTypes.shape() },
      },
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
