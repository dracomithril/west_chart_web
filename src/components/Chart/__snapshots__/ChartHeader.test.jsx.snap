// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChartHeader/> renders without crashing ChartHeader 1`] = `
<ChartHeader
  classes={
    Object {
      "button": "ChartHeader-button-1",
      "container": "ChartHeader-container-2",
    }
  }
  errorDays={Array []}
/>
`;
